### DISCUSSION TITLE
FR: add a "recreate" attribute to cache tag

### DISCUSSION THREAD

**Discussion by hiasl** [craftcms/cms#15005] [Ideas]
Hi!

We currently have a scenario where we want to recreate/refresh the content of a `{% cache %}` tag without losing its cache first. With losing the cache I mean either by timing it out or by clearing the cache manually. 

The idea was to call a page's URL with a `refresh=true` GET param or even better a certain http header and the `{% cache %}` tag would not check if it has a cached content but simple enforce cache recreation. It's important to understand that this can not be achieved with the` if/unless` parameters, as they disable the cache tag completely, meaning that no cached content is stored.

Our idea would be a simple `recreate` param on the `{% cache %}` tag which causes this behavior.

If you want to see our dirty workaround, please read this: https://craftcms.stackexchange.com/questions/41717/recreate-cache-content-without-clearing-first

Any feedback to our FR is appreciated.

Matthias

**Comment by brandonkelly** [comment]
> Our idea would be a simple `recreate` param on the `{% cache %}` tag which causes this behavior.

What exactly do you mean by that? Are you thinking that would be set to a condition?

```twig
{% cache recreate craft.app.request.getQueryParam('refresh') == 'true' %}
```

Not sure I love that syntax. `recreate if`/`recreate when` would read a little better, but that would be confusing alongside the `if` param.

Maybe this makes more sense as a `regenerateAllCaches` setting on the `TemplateCaches` service, which you could set from `config/app.web.php`:

```php
return [
    'components' => [
        'templateCaches' => [
            'regenerateAllCaches' => fn() => Craft::$app->request->getBodyParam('refresh') === 'true',
        ],
    ],
];
```

**Reply by hiasl** [reply]
yes, you get the idea! I want to put a condition on the cache tag, but preferably use a request header instead instead of a GET, because unless set manually, it would become part of the cache key.

Regarding the syntax: I understand that you don't like `recreate if`, but what about `recreateif`, `recreate-if` or `refresh-if`?

your suggestion about a `regenerateAllCaches` setting on the `TemplateCaches` service ... does that already work? and is this specific to URL currently requested or would it purge all template caches? 

In general I preferred a solution that always works instead of having to configure it first, but if it does the job ... it's at least a good workaround.

**Reply by brandonkelly** [reply]
> your suggestion about a regenerateAllCaches setting on the TemplateCaches service ... does that already work? and is this specific to URL currently requested or would it purge all template caches?

No, just an alternate possible solution. Going with your use case, if you’re passing a header that is meant to get all caches to regenerate, it seems like it would make more sense to tell Craft to regenerate all caches centrally via the app config, rather than on each individual `{% cache %}` tag.

**Comment by hiasl** [comment]
I'm not sure if I understand you: Would the suggested solution in `app.web.php` only regenerate the templates caches of the requested URL (requested with the additional header), or would it actually clear all template caches (like a cache purge), also of other URLs? Because I want to refresh my URLs one by one to keep the load low.

My preference would still be an option on the `cache` tag for 2 reasons:
- Others would see it in the cache tag documentation and could adopt it more easily
- It gives your more control on which `cache` tags to refresh

But at the end, both solutions do the job.

**Reply by brandonkelly** [reply]
It would only refresh the caches for the current request, if the header is present. If you look at the config example, it’s defining the `regenerateAllCaches` value dynamically based on the presence of a body param – but that could have been a header instead.

**Reply by hiasl** [reply]
ok
